Для анализа был выбран файл, содержащий текст стихотворения С. Есенина "Письмо к женщине" и скачанный с сайта deterra.ru.

In [1]:
#этап 1: загружаем файл и выполняем его базовую обработку

text = []
start_poem = False #переменная, означающая начало текста стихотворения
num_lines = 0 #общее количество строк в файле

with open('pismo-k-zhenshchine.txt', mode='r', encoding='utf-8') as infile:
    for line in infile:
        line = line.strip() #удаляем пробелы в конце строк
        num_lines += 1

        if len(line) == 0 or line == "………..": #удаляем пустые строки
            continue

        if line == "⚐":
            start_poem = True
            continue

        if not start_poem: #пропускаем ненужные строки
            continue

        words = line.split() #список слов в 1 строке
        text.append(words)

print(f'Количество строк в исходном файле: {num_lines}')
print('\nТекст (первые 10 строк):')
for i, line in enumerate(text[:10]):
    print(f'{i+1}: {line}')

Количество строк в исходном файле: 124

Текст (первые 10 строк):
1: ['Вы', 'помните,']
2: ['Вы', 'всё,', 'конечно,', 'помните,']
3: ['Как', 'я', 'стоял,']
4: ['Приблизившись', 'к', 'стене,']
5: ['Взволнованно', 'ходили', 'вы', 'по', 'комнате']
6: ['И', 'что-то', 'резкое']
7: ['В', 'лицо', 'бросали', 'мне.']
8: ['Вы', 'говорили:']
9: ['Нам', 'пора', 'расстаться,']
10: ['Что', 'вас', 'измучила']


В исходном файле содержится 124 строки. Теперь приступим непосредственно к анализу текста стихотворения.

In [2]:
#этап 2: очищаем текст от знаков препинания и вычисляем количество предложений в стихотворении

num_sentences = 0 #количество предложений в стихотворении
end_sentence = '!?.…' #только завершающие знаки препинания
punctuation = '",;:' + end_sentence #все знаки препинания

cleaned_text = []

for line in text:
    cleaned_line = []
    for word in line:
        if word == '—':
            continue
        if word[-1] in end_sentence:
            num_sentences += 1
        cleaned_word = word.strip(punctuation)
        cleaned_line.append(cleaned_word)
    cleaned_text.append(cleaned_line)

print(f'Количество предложений в стихотворении: {num_sentences}')
print('\nОчищенный текст (первые 10 строк):')
for i, line in enumerate(cleaned_text[:10]):
    print(f'{i+1}: {line}')

Количество предложений в стихотворении: 35

Очищенный текст (первые 10 строк):
1: ['Вы', 'помните']
2: ['Вы', 'всё', 'конечно', 'помните']
3: ['Как', 'я', 'стоял']
4: ['Приблизившись', 'к', 'стене']
5: ['Взволнованно', 'ходили', 'вы', 'по', 'комнате']
6: ['И', 'что-то', 'резкое']
7: ['В', 'лицо', 'бросали', 'мне']
8: ['Вы', 'говорили']
9: ['Нам', 'пора', 'расстаться']
10: ['Что', 'вас', 'измучила']


In [3]:
#этап 3: вычисляем количество слов в стихотворении

num_words = 0 #количество слов в стихотворении

for line in cleaned_text:
    for word in line:
        num_words += 1

print(f'Количество слов в стихотворении: {num_words}')

Количество слов в стихотворении: 376


In [4]:
#этап 4: создаем маску для слов, которые не являются предлогами, союзами или местоимениями

def meaningful_word_mask(word):
    if len(word) > 3: #предположим, что слово не является предлогом, союзом или местоимением, если его длина превышает 3 символа
        return True
    else:
        return False

In [5]:
#этап 5: определяем 10 самых частотных слов (включая предлоги, союзы и местоимения)

words = {} #словарь с уникальными словами и частотой их встречаемости в тексте
#{слово 1: частота слова 1, слово 2: частота слова 2, ...}

for line in cleaned_text:
    for word in line:
        word = word.lower() #приводим слова к нижнему регистру

        if word not in words:
            words[word] = 1
        else:
            words[word] += 1

sorted_words = sorted(words.items(), key=lambda x: x[1], reverse=True) #cортируем слова по убыванию их частоты
print('10 самых частотных слов:')
for i, (word, count) in enumerate(sorted_words[:10]):
    print(f'{i+1}. {word} — {count} раз')

10 самых частотных слов:
1. я — 18 раз
2. в — 17 раз
3. не — 17 раз
4. и — 15 раз
5. вы — 10 раз
6. что — 10 раз
7. вас — 6 раз
8. с — 6 раз
9. как — 4 раз
10. вам — 4 раз


In [12]:
#этап 6: вычисляем частоту введенного пользователем слова

request = input('Введите слово: ')
request = request.lower() #приводим слово к нижнему регистру

if request in words:
    print(f'Слово "{request}" встречается в тексте {words[request]} раз')
else:
    print(f'Слово "{request}" в стихотворении не встречается')

Введите слово:  Любовь


Слово "любовь" в стихотворении не встречается


In [7]:
#этап 7: определяем 10 самых частотных содержательных слов (исключая предлоги, союзы и местоимения)

meaningful_words = {} #словарь с уникальными содержательными словами и частотой их встречаемости в тексте

for word, count in words.items():
    if meaningful_word_mask(word): #применяем маску
        meaningful_words[word] = count

sorted_meaningful_words = sorted(meaningful_words.items(), key=lambda x: x[1], reverse=True) #сортируем слова по убыванию их частоты
print('10 самых частотных содержательных слов:')
for i, (word, count) in enumerate(sorted_meaningful_words[:10]):
    print(f'{i+1}. {word} — {count} раз')

10 самых частотных содержательных слов:
1. любимая — 3 раз
2. знали — 3 раз
3. теперь — 3 раз
4. помните — 2 раз
5. пора — 2 раз
6. сплошном — 2 раз
7. дыму — 2 раз
8. развороченном — 2 раз
9. бурей — 2 раз
10. быте — 2 раз


In [8]:
#этап 8: определяем среднюю длину слов в стихотворении

cleaned_text_str = ' '.join(' '.join(line) for line in cleaned_text)
cleaned_text_list = cleaned_text_str.split()

sum_word_length = 0
for word in cleaned_text_list:
    sum_word_length += len(word)

av_word_length = sum_word_length / num_words
print(f'Средняя длина слов в стихотворении: {round(av_word_length, 2)} букв')

Средняя длина слов в стихотворении: 4.39 букв


In [9]:
#этап 9: определяем среднюю длину предложений в стихотворении

text_str = ' '.join(' '.join(line) for line in text)

sent_lengths_list = [] #список с длинами каждого предложения
sent_length = 0
for symbol in text_str:
    if symbol not in end_sentence:
        sent_length += 1
    else:
        sent_length += 1
        sent_lengths_list.append(sent_length)
        sent_length = 0

for i in range(1, len(sent_lengths_list)): #удаляем учет пробелов
    sent_lengths_list[i] -= 1

sum_sent_length = sum(sent_lengths_list)
av_sent_length = sum_sent_length / num_sentences
print(f'Средняя длина предложений в стихотворении: {round(av_sent_length, 2)} символов')

Средняя длина предложений в стихотворении: 59.57 символов


In [10]:
#этап 10: сравниваем лексику, используемую в начале и в конце стихотворения

N = 5 #количество рассматриваемых строк
beginning = []
ending = []

for line in cleaned_text[:N]:
    beginning.extend(line)

for line in cleaned_text[-N:]:
    ending.extend(line)

beginning_set = set(word.lower() for word in beginning)
ending_set = set(word.lower() for word in ending)

common_words = beginning_set & ending_set #уникальные слова, общие для начала и конца стихотворения
unique_to_beginning = beginning_set - ending_set
unique_to_ending = ending_set - beginning_set

print(f'Количество общих слов начала и конца стихотворения: {len(common_words)}')
if len(common_words) > 0:
    print(f'Общие слова: {", ".join(common_words)}')

print(f'\nКоличество уникальных слов, встречающихся только в начале стихотворения: {len(unique_to_beginning)}')

print(f'\nКоличество уникальных слов, встречающихся только в конце стихотворения: {len(unique_to_ending)}')

Количество общих слов начала и конца стихотворения: 0

Количество уникальных слов, встречающихся только в начале стихотворения: 14

Количество уникальных слов, встречающихся только в конце стихотворения: 13


Наконец, запишем результаты анализа в файл "results.txt".

In [11]:
#этап 11: сохраняем результаты в файл

with open('results.txt', mode = 'w', encoding = 'utf-8') as outfile:
    print(f'Количество строк в исходном файле: {num_lines}', file=outfile)

    print(' ', file=outfile)

    print(f'Количество предложений в стихотворении: {num_sentences}', file=outfile)

    print(' ', file=outfile)

    print(f'Количество слов в стихотворении: {num_words}', file=outfile)

    print(' ', file=outfile)

    print('10 самых частотных слов:', file=outfile)
    for i, (word, count) in enumerate(sorted_words[:10]):
        print(f'{i+1}. {word} — {count} раз', file=outfile)

    print(' ', file=outfile)

    if request in words:
        print(f'Слово "{request}" встречается в тексте {words[request]} раз', file=outfile)
    else:
        print(f'Слово "{request}" в стихотворении не встречается', file=outfile)

    print(' ', file=outfile)

    print('10 самых частотных содержательных слов:', file=outfile)
    for i, (word, count) in enumerate(sorted_meaningful_words[:10]):
        print(f'{i+1}. {word} — {count} раз', file=outfile)

    print(' ', file=outfile)

    print(f'Средняя длина слов в стихотворении: {round(av_word_length, 2)} букв', file=outfile)

    print(' ', file=outfile)

    print(f'Средняя длина предложений в стихотворении: {round(av_sent_length, 2)} символов', file=outfile)

    print(' ', file=outfile)

    print(f'Количество общих слов начала и конца стихотворения: {len(common_words)}', file=outfile)
    if len(common_words) > 0:
        print(f'Общие слова: {", ".join(common_words)}', file=outfile)
    print(f'Количество уникальных слов, встречающихся только в начале стихотворения: {len(unique_to_beginning)}', file=outfile)
    print(f'Количество уникальных слов, встречающихся только в конце стихотворения: {len(unique_to_ending)}', file=outfile)